### DISCUSSION TITLE
Ability to opt-out of `web-vitals` package that's currently bundled with Next.js

### DISCUSSION THREAD

**Discussion by dmitrizzle** [vercel/next.js#15414] [Ideas]
# Feature request

Currently, Next.js adds `web-vitals` package and a file that handles it -- into the production bundle. Though it may be useful in many cases, my project does not rely on these metrics. It adds an extra KB of code and whatever the parsing resources it needs from the user's machine. This isn't much, but I think that in line with Next.js's telemetry opt-out we should be able to opt out of this feature as well.

## Describe the solution you'd like

Add a flag/setting in next.config.js to opt-out of `web-vitals`.

## Describe alternatives you've considered

The only thing to do here, as I understand, other than setting up the above flag is to fork Next.js.

## Additional context

https://github.com/vercel/next.js/discussions/15296

**Comment by housseindjirdeh** [comment]
Although it's tiny, I agree it's an additional cost if you're not using it to measure metrics at all.

I wouldn't mind if we keep it included by default and have a flag to opt-out. @timneutkens do you agree?

**Comment by timneutkens** [comment]
There's likely a better approach possible here where we remove the web vitals package during compilation when reporting is not set up

**Comment by RodrigoTomeES** [comment]
Hi,

Even in 2022 there is this problem, is there any way to fix it? Or is there any workaround?

Thank you very much!!

**Comment by dmitrizzle** [comment]
Looks like this is only in the interest of the users and not maintainers, so it's probably up to someone outside Vercel to make a PR and hope they'll allow it.

**Comment by MattCCC** [comment]
Bump, it feels like it's an unnecessary bloat if one does not use it.

**Reply by timneutkens** [reply]
This request was kept in mind while building App Router and it has been fixed there. We can't remove it from Pages Router as it'd break existing applications.

**Reply by yunsii** [reply]
But I think provide a setting to opt-out of web-vitals is possible? And PageSpeed throw a DIAGNOSTIC:
![image](https://github.com/vercel/next.js/assets/18096089/a4582783-96fd-4531-aa96-90d2e55d13e0)
![image](https://github.com/vercel/next.js/assets/18096089/a4fab11a-635b-4c48-9894-3b7ddbfd1e9a)

After further investigation, I found it come from `web-vitals`.

**Reply by RodrigoTomeES** [reply]
> This request was kept in mind while building App Router and it has been fixed there. We can't remove it from Pages Router as it'd break existing applications.

Why could not be added an option to opt it out? This not break existing apps

**Reply by varghesethomase** [reply]
@timneutkens I would appreciate it if you could point me to the documentation on how to turn it off in app router. I am using the app router and could see that its still  a part of the production build.
<img width="1160" alt="image" src="https://github.com/vercel/next.js/assets/5429756/1db8c50f-ebad-4d6f-af44-03a89b47368d">
<img width="352" alt="image" src="https://github.com/vercel/next.js/assets/5429756/b03c8f88-4985-4f9e-97bd-aa7104f4dc1d">

**Reply by timneutkens** [reply]
You're looking at files for pages router. Here's the analysis for App Router only:

<img width="2060" alt="CleanShot 2023-12-04 at 11 25 33@2x" src="https://github.com/vercel/next.js/assets/6324199/bc14f9b9-91f6-492e-a517-4809d839f0c8">

**Comment by MattCCC** [comment]
It is really annoying that Next.js increases bundle size with useless functionality. It is completely unnecessary in most of the projects. Web vitals should be opt in.

**Comment by sebastiangrebe** [comment]
Any update on this?

**Reply by timneutkens** [reply]
I've already replied on this. See https://github.com/vercel/next.js/discussions/15414#discussioncomment-6527735
